src/core/namespace.c

```diff
@@ -287,10 +288,18 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                // mount() always follow symlinks, so we follow symlinks during the stat() as well
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return r;
+                if (!S_ISDIR(target.st_mode))
+                        what = "/dev/null";
+                else {
+                        what = "/run/systemd/inaccessible";
+                }
                 break;
```

Comment:
hmm, instead of using /dev/null as replacement for block and char devices we could also create block/char devices with major/minor of 0 for each, which I think will always result in ENODEV when accessed, which is probably pretty good semantics. But this would need verification...